### PR TITLE
Migrate Q.finite and Q.infinite to new predicate design

### PR DESCRIPTION
**Pull request by JSS95**
#### References to other Issues or PRs

Succeeds #20656

#### Brief description of what is fixed or changed
This PR migrates the part of old-design predicates to new design.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/assumptions/ask_generated.py [hunk 1]
```diff
@@ -99,27 +99,29 @@ def get_known_facts_cnf():
         Q.complex | ~Q.imaginary,
         Q.complex | ~Q.real,
         Q.complex | ~Q.transcendental,
+        Q.extended_real | ~Q.infinite,
```
Comment (oscarbenjamin) [inline sympy/assumptions/ask_generated.py:102]: Why has this file changed?

Comment (JSS95) [reply sympy/assumptions/ask_generated.py:102]: It seems that change of class structure affects the order of this dictionary. `Q.infinite` was `UndefinedPredicate` before migration, but now it's `InfinitePredicate`.

Comment (oscarbenjamin) [reply sympy/assumptions/ask_generated.py:102]: As an aside in the old assumptions we do not have this rule and it is inconsistent with zoo.

It is possible to be both infinite and not extended real e.g.:
```python
In [33]: x = Symbol('x', infinite=True, extended_real=False)

In [34]: x.is_infinite
Out[34]: True

In [35]: x.is_extended_real
Out[35]: False

In [36]: zoo.is_infinite
Out[36]: True

In [37]: zoo.is_extended_real
Out[37]: False
```
Not for this PR but that should be fixed.

File: sympy/assumptions/handlers/calculus.py [hunk 2]
```diff
@@ -3,232 +3,205 @@
...
+        return True
+    if (abs(expr.base) >= 1) == True and ask(Q.negative(expr.exp), assumptions):
+        return True
+    if (abs(expr.base) >= 1) == True and exp_bounded is False:
+        return False
+    return None
+
+@FinitePredicate.register_many(exp, log)
+def _(expr, assumptions):
+    return ask(Q.finite(expr.args[0]), assumptions)
```
Comment (oscarbenjamin) [inline sympy/assumptions/handlers/calculus.py:198]: This isn't valid for `log` is it?

Zero is finite but `log(0)` is not.

Comment (oscarbenjamin) [reply sympy/assumptions/handlers/calculus.py:198]: If that is just preserving existing behaviour then I don't suggest to change it in this PR.

**Comment by oscarbenjamin** [comment]
I had a query about `log` but if that's just preserving existing behaviour then this looks good to me. (I'll let you merge)

**Comment by JSS95** [comment]
@oscarbenjamin 
I will make issues for these matters.

**Comment by JSS95** [comment]
@oscarbenjamin 
I will open an issue for these matters.

**Pull request merged**